### PR TITLE
Feat/1561 form outgoing next gen xms requests

### LINKED ISSUE #1561: As a developer, I need to be able to form outgoing NextGenXMS requests
**Description:**
_Provide a brief background and justification for this issue_

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] Ensure user is redirected back to our application (`/v1/loginXMS/` route) after authentication with NextGenXMS
- [x] Direct user to NextGenXMS with appropriate metadata
- [x] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary
- [x] [`openid-connect.md`](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Technical-Documentation/openid-connect.md) is updated to match these new requests for NextGenXMS.

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

 - [x] Perform basic tests on OpenID endpoint via curl/bash/etc.
 - [x] Create XMS structure with associated environment variables, JWT, payload items, etc. like in #1136
- [x] Package and encode request for NextGenXMS OpenID endpoint
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #1561 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Tabletop

- [x] Ready for Tabletop (Reviewer check this)
- [x] Tabletop scheduled (Assignee check this)
- [x] Tabletop approval 

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x]  Ensure user is redirected back to our application (/v1/loginXMS/ route) after authentication with NextGenXMS
+ [x]  Direct user to NextGenXMS with appropriate metadata
+ [x]  Testing Checklist has been run and all tests pass
+ [x]  README is updated, if necessary
+ [x]  [openid-connect.md](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Technical-Documentation/openid-connect.md) is updated to match these new requests for NextGenXMS.
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/settings/common.py [hunk 1]
```diff
@@ -368,6 +370,32 @@ class Common(Configuration):
...
+        'http://idmanagement.gov/ns/assurance/ial/1'
+    )
+    XMS_AUTHORIZATION_ENDPOINT = 'https://int.xms.hhs.gov/ABaaS/oidc/authorize'
+
+    XMS_CLIENT_ASSERTION_TYPE = ''
+    XMS_CLIENT_ID = 'U2Xzpq6_ZA_5M~Z'
+    XMS_ISSUER = "https://int.xms.hhs.gov/"
+    XMS_JWKS_ENDPOINT = "https://int.xms.hhs.gov/ABaaS/oidc/certificates"
+    # JWT_KEY must be set, there is no functional default.
+    XMS_JWT_KEY = 'n5i8ZxW.I8N68iobatUBR~ClLkc7t5q2~.oZXkx.N7.L-jvJuo'
```
Comment (riatzukiza) [inline tdrs-backend/tdpservice/settings/common.py:392]: this seems like it should be an environment variable?

File: tdrs-backend/tdpservice/settings/common.py [hunk 2]
```diff
@@ -368,6 +370,32 @@ class Common(Configuration):
...
+    "jwks_uri":"https://int.xms.hhs.gov/ABaaS/oidc/certificates"
+    '''
+    XMS_ACR_VALUES = os.getenv(
+        'ACR_VALUES',
+        'http://idmanagement.gov/ns/assurance/ial/1'
+    )
+    XMS_AUTHORIZATION_ENDPOINT = 'https://int.xms.hhs.gov/ABaaS/oidc/authorize'
+
+    XMS_CLIENT_ASSERTION_TYPE = ''
+    XMS_CLIENT_ID = 'U2Xzpq6_ZA_5M~Z'
```
Comment (riatzukiza) [inline tdrs-backend/tdpservice/settings/common.py:388]: should probably be an environment variable

**Review by riatzukiza — CHANGES_REQUESTED** [review]

File: tdrs-backend/tdpservice/settings/common.py (hunk 2 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/settings/common.py:388]: Correct, will do the change

File: tdrs-backend/tdpservice/settings/common.py (hunk 1 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/settings/common.py:392]: Correct, will do the change

File: tdrs-backend/tdpservice/users/api/login.py [hunk 3]
```diff
@@ -325,6 +325,53 @@ def verify_email(self, user):
...
+
+    def get_auth_options(self, access_token, sub):
+        """Add specific auth properties for the CustomAuthentication handler."""
+        # TODO: Refactor to XMS login
+        pass
+
+    def verify_email(self, user):
+        """Handle user email exception to disallow ACF staff to utilize non-AMS authentication."""
+        # TODO: Refactor to XMS login
+        pass
```
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login.py:349]:
```suggestion

```

Not needed for XMS, the inherited function from TokenAuthorizationOIDC should be sufficient.

File: docs/Technical-Documentation/openid-connect.md [hunk 4]
```diff
@@ -2,37 +2,38 @@
...
 ![generic OIDC flow](https://miro.medium.com/max/1080/1*quwFs1fFCvTvLT80e_QHVA.png)
 
-## Step 1: Authentication & Authorization Request
+## XMS
+
+### Step 1: Authentication & Authorization Request
 * User clicks Sign In button on site OR navigates to Django Admin unauthenticated
-* Call is made to /login/oidc - which initiates [LoginRedirectOIDC view](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/users/api/login_redirect_oidc.py)
-* Backend constructs a URL encoded query string for redirect to Login.gov, which consists of:
+* Call is made to ../login/xms - which initiates [LoginRedirectOIDC view](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/users/api/login_redirect_oidc.py)
```
Comment (abottoms-coder) [inline docs/Technical-Documentation/openid-connect.md:9]:
```suggestion
* Call is made to `/login/xms` - which initiates [LoginRedirectOIDC view](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/users/api/login_redirect_oidc.py)
```

File: docs/Technical-Documentation/openid-connect.md [hunk 5]
```diff
@@ -2,37 +2,38 @@
...
+* In our setup, this all happens on the XMS which redirects user to Login.gov side
 * Once user has logged in successfully, Login.gov redirects them back to the `redirect_uri` we specified above
-* NOTE: Any `redirect_uri` used must be defined in Login.gov or it will be rejected
+* NOTE: Any `redirect_uri` used must be defined in XMS or it will be rejected
 
-## Step 3: Authorization Code
-* User is redirected to /login - which correlates to the [TokenAuthorizationOIDC view](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/users/api/login.py)
-* The Authorization Code (`code`), Nonce and State are extracted from the query parameters supplied to the GET request
+### Step 3: Authorization Code
+* User is redirected to ../oidc/xms - which correlates to the [TokenAuthorizationOIDC view](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/users/api/login.py)
```
Comment (abottoms-coder) [inline docs/Technical-Documentation/openid-connect.md:25]:
```suggestion
* User is redirected to `/oidc/xms` - which correlates to the [TokenAuthorizationOIDC view](../../tdrs-backend/tdpservice/users/api/login.py)
```

**Review by abottoms-coder — CHANGES_REQUESTED** [review]
Had some minor requests but otherwise this looks good to me.

**Review by abottoms-coder — APPROVED** [review]

**Review by jorgegonzalez — APPROVED** [review]
LGTM

**Comment by ADPennington** [comment]
@raftmsohani below are the results after deploying to [qasp](https://tdp-frontend-qasp.app.cloud.gov/) environment. can you let me know if this is expected behavior, since this is only the first part?

1. click login.gov button

takes me here: https://int.xms.hhs.gov/ABaaS/oidc/authorize?acr_values=http://idmanagement.gov/ns/assurance/ial/1&client_id=LGpyp11pK2vACe4&nonce=206da6404a2e886948e4d4b225ca41d929a7d007a01042d1e9cf284c0ffc35cf&redirect_uri=https://tdp-backend-qasp.app.cloud.gov/v1/oidc/xms&response_type=code&state=00ac3860f401a4a5535cc41444344d36dd15356887dea8f201dc7bd36e955d54&scope=openid+email




From here, I tried each button. 

- xms button, i authenticate and am taken to the landing page: https://int.xms.hhs.gov/xms-app/landing
1561p3

- login.gov button takes me to the official site, which has no record of my account (which is expected since the account was created in
1561p2
 the login.gov sandbox)

**Comment by raftmsohani** [comment]
> @raftmsohani below are the results after deploying to [qasp](https://tdp-frontend-qasp.app.cloud.gov/) environment. can you let me know if this is expected behavior, since this is only the first part?
> 
> 1. click login.gov button
> 
> takes me here: https://int.xms.hhs.gov/ABaaS/oidc/authorize?acr_values=http://idmanagement.gov/ns/assurance/ial/1&client_id=LGpyp11pK2vACe4&nonce=206da6404a2e886948e4d4b225ca41d929a7d007a01042d1e9cf284c0ffc35cf&redirect_uri=https://tdp-backend-qasp.app.cloud.gov/v1/oidc/xms&response_type=code&state=00ac3860f401a4a5535cc41444344d36dd15356887dea8f201dc7bd36e955d54&scope=openid+email
> 
> 
> 
> From here, I tried each button.
> 
> * xms button, i authenticate and am taken to the landing page: https://int.xms.hhs.gov/xms-app/landing
>   1561p3
> * login.gov button takes me to the official site, which has no record of my account (which is expected since the account was created in
>   1561p2
>   the login.gov sandbox)

@ADPennington : The first part is not what I had expected, but from what I know the response page has changed. It should give a warning page and then the two buttons. But then you have the buttons, and can authenticate! So that part is correct.

**Comment by ADPennington** [comment]
@raftmsohani @andrew-jameson just logged in to sandbox and the results were the same as above 🆙 (first 2 screenshots). not making it back to frontend.

**Comment by raftmsohani** [comment]
@ADPennington : you will need XMS_CLIENT_ID and XMS_JWT_KEY in env. @andrew-jameson : Could you request these for Alex?

**Comment by andrew-jameson** [comment]
> @ADPennington : you will need XMS_CLIENT_ID and XMS_JWT_KEY in env. @andrew-jameson : Could you request these for Alex?

Set these variables and restarted and getting redirected:
image

**Review by ADPennington — APPROVED** [review]
tested with 2 non-acf accounts after keys were changed and can confirm that I also get the handled error. thanks @raftmsohani and @andrew-jameson 🚀

**Pull request merged**